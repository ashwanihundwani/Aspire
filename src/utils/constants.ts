import { windowWidth, windowHeight } from "./utils";
import { Platform } from "react-native";

// Tab Name Constants
export const Tabs = { 
    home: "Home", 
    debit: "Debit Card", 
    payments: "Payments",
    credit: "Credit",
    account: "Profile"
} as const;

// Stack Route/Name Constants
export const StackScreens = { 
    debitCardTab: "Tab", 
    weeklySpentLimit: "WeeklySpendingLimit", 
    
} as const;

// Font Name Constants
export const FontNames = { 
    Avenir: "AvenirNext"  
} as const;

export const AppColors = {
    lightGray: "#DDDDDD",
    lightGreen: "#01D167",
    navigationBackColor: "#0C365A",
    navigationTintColor: "#FFFFFF",
    whiteColor: "#FFFFFF",
    darkTextColor: "#22222266",
    descriptionColor: "#222222",
    spentLimitText: "#22222233",
    separatorColor: "#DDDDDD",
    lightGreenWithOpacity: 'rgba(1,209,103, 0.07)',
    debitCardTitleColor: "#25345F",
    disabledSaveButton: "#EEEEEE"
}

export const CommonString = {
    currencySign: "S$",
    dollarSign: "$",
    switchSuffix:"_Switch"
}

export const DebitCardString = {
    headerTitle: "Debit Card",
    expiry: "Thru: ",
    cvv: "CVV: ",
    showCard: "Show card number",
    hideCard: "Hide card number",
    availableBalance: "Available Balance",
    spendLimitLabel: "Debit card spending limit",
    cardInfo: {
        cardNumberSplitter:"-",
        hiddenCVV: "* * *",
        dotString: '\u2022\u2022\u2022\u2022'
    }
}

export const WeeklySpendingString = {
    headerTitle: "Spending Limit",
    descriptionMessage: "Set a weekly debit card spending limit",
    infoMessage: "Here weekly means the last 7 days - not the calendar week", 
    saveButton: "Save",
    buttonTitle1: "S$ 5,000",
    buttonTitle2: "S$ 10,000",
    buttonTitle3: "S$ 20,000",
}

export const DebitCardOptions = {
    Titles : {
        topup: "Top-up account",
        weeklyspent: "Weekly spending limit",
        freezeCard: "Freeze card",
        newCard: "Get a new card",
        deactivated: "Deactivated cards"
    },
    Descriptions: {
        topup: "Deposit money to your account to use with card",
        weeklyspent: "You havenâ€™t set any spending limit on card",
        freezeCard: "Your debit card is currently active",
        newCard: "This deactivates your current debit card",
        deactivated: "Your previously deactivated cards"
    },
}

export const layouting = {
    // Standard margin calculation to accomodate various screen sizes accross devices.
    spacingFactor: 0.058 * windowWidth(),
}

export const testIDs = {
    showHideButton: "DebitCard.CardInfo.showHide",
    cardOption: "DebitCard.CardOption",
    spendProgress:"DebitCard.SpendProgress"
}

  