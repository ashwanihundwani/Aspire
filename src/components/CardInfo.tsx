import React, { useState, useEffect } from 'react'
import { View, StyleSheet, Image, Text, TouchableOpacity, Platform } from 'react-native'
import { AppColors, DebitCardString, layouting, testIDs } from '../utils/constants'
import { windowWidth, avenirFont, FontStyle } from '../utils/utils'

import Images from '../../src/assets/images';
import SpendLimitProgress from './SpendLimitProgress'

const { cardInfo } = DebitCardString

const spacingFactor: number = layouting.spacingFactor
const screenWidth: number = windowWidth()
const screenMargin: number = spacingFactor * 2
const cardNumberSplitter: string = cardInfo.cardNumberSplitter
const dotString: string = cardInfo.dotString
const hiddenCVV: string = cardInfo.hiddenCVV

interface CardInfoProps {
    holderName: string,
    cardNumber: string,
    expiry: string,
    CVV: string,
    weeklySpentLimit: string,
    amountSpent: string,
    showProgress: boolean
}

const CardInfo: React.FC<CardInfoProps> = (props: CardInfoProps) => {

    const [showInfo, setShowInfo] = useState<boolean>(true)
    const [cardNumber, setCardNumber] = useState<Array<string>>([])

    const cardNumberTokens = (cardNumber: string) => {
        const cardNumberTokens: string[] = cardNumber.split(cardNumberSplitter)
        let cardNumberDisplay: string[] = []
        cardNumberTokens.forEach((token) => {
            cardNumberDisplay.push(token);
        })
        setCardNumber(cardNumberDisplay)
    }

    const cardNumberTokensHidden = (cardNumber: string) => {
        const cardNumberTokens: string[] = cardNumber.split("-")
        let cardNumberDisplay: string[] = []
        let index: number = 0
        cardNumberTokens.forEach((token) => {
            if (index < cardNumberTokens.length - 1) {
                cardNumberDisplay.push(dotString)
            }
            else {
                cardNumberDisplay.push(token)
            }
            index++
        })
        setCardNumber(cardNumberDisplay)
    }

    useEffect(()=> {
        if(showInfo) {
            cardNumberTokens(props.cardNumber)
        }
        else {
            cardNumberTokensHidden(props.cardNumber)
        }
    }, [showInfo])

    return <View style={styles.container}>
        <TouchableOpacity
            testID={testIDs.showHideButton}
            style={styles.showButton}
            onPress={() => {
                setShowInfo(!showInfo)
            }}>
            <View style={styles.showButtonContent}>
                <Image source={showInfo ? Images.cardInfo.showIcon : Images.cardInfo.hideIcon} />
                <Text style={styles.showButtonText}>{showInfo ? DebitCardString.showCard : DebitCardString.hideCard}</Text>
            </View>
        </TouchableOpacity>
        <View style={styles.card}>
            <Image style={styles.aspireLogo} source={Images.cardInfo.aspireLogo} />
            <Text style={styles.holderName}>{props.holderName}</Text>
            <View style={styles.cardNumberContainer}>
                <Text style={styles.cardNumber}>{cardNumber[0]}</Text>
                <Text style={styles.cardNumber}>{cardNumber[1]}</Text>
                <Text style={styles.cardNumber}>{cardNumber[2]}</Text>
                <Text style={styles.cardNumber}>{cardNumber[3]}</Text>
            </View>
            <View style={styles.expiryCVVContainer}>
                <Text style={styles.expiry}>{DebitCardString.expiry + props.expiry}</Text>
                <Text style={styles.cvv}>{DebitCardString.cvv}<Text style={showInfo ? styles.cvvValue : styles.hiddenCVV}>{(showInfo ? props.CVV : hiddenCVV)}</Text></Text>
            </View>
            <Image style={styles.visaLogo} source={Images.cardInfo.visa} />

        </View>
        {props.showProgress && <SpendLimitProgress spent={props.amountSpent} limit={props.weeklySpentLimit} />}
    </View>
}

export default React.memo<CardInfoProps>(CardInfo);

const styles = StyleSheet.create({
    container: {
        width: "100%",
        marginTop: 34,
        marginBottom: 34
    },
    card: {
        marginTop: 0,
        alignSelf: "center",
        width: screenWidth - screenMargin,
        aspectRatio: 1.66,
        backgroundColor: AppColors.lightGreen,
        borderRadius: 14,
    },
    aspireLogo: {
        alignSelf: "flex-end",
        width: 74,
        height: 21,
        marginRight: spacingFactor,
        marginTop: spacingFactor
    },
    holderName: {
        color: AppColors.whiteColor,
        marginLeft: spacingFactor,
        marginTop: Platform.OS === 'ios' ? spacingFactor : spacingFactor - 10,
        fontFamily: avenirFont(FontStyle.Bold),
        fontSize: 22
    },
    cardNumberContainer: {
        flexDirection:"row",
        marginLeft: spacingFactor,
        marginTop: Platform.OS === 'ios' ? spacingFactor : spacingFactor - 5,
    },
    cardNumber: {
        marginRight:spacingFactor,
        letterSpacing: 3.46,
        color: AppColors.whiteColor,
        fontFamily: avenirFont(FontStyle.DemiBold),
        fontSize: 14
    },
    expiryCVVContainer: {
        flexDirection: "row",
        justifyContent: "flex-start",
        marginTop: Platform.OS === 'ios' ? 15 : 8
    },
    expiry: {
        color: AppColors.whiteColor,
        marginLeft: spacingFactor,
        fontFamily: avenirFont(FontStyle.DemiBold),
        fontSize: 13
    },
    cvv: {
        fontSize: 13,
        color: AppColors.whiteColor,
        marginLeft: 32,
        fontFamily: avenirFont(FontStyle.DemiBold),
    },
    cvvValue: {
        fontSize: 13,
        color: AppColors.whiteColor,
        marginLeft: 32,
        fontFamily: avenirFont(FontStyle.DemiBold),
    },
    hiddenCVV: {
        fontSize: 13,
        color: AppColors.whiteColor,
        marginLeft: 32,
        fontFamily: avenirFont(FontStyle.DemiBold),
    },
    visaLogo: {
        alignSelf: "flex-end",
        width: 59,
        height: 20,
        marginRight: spacingFactor,
        
    },
    showButton: {
        marginBottom: -10,
        borderTopLeftRadius: 6,
        borderTopRightRadius: 6,
        alignSelf: "flex-end",
        width: 151,
        height: 38,
        marginRight: spacingFactor,
        backgroundColor: AppColors.whiteColor
    },
    showButtonContent: {
        marginTop: 5,
        flexDirection: "row",
        alignSelf: "center"
    },
    showButtonText: {
        marginLeft: 6,
        color: AppColors.lightGreen,
        fontFamily: avenirFont(FontStyle.DemiBold),
        fontSize: 12,
    }
})
