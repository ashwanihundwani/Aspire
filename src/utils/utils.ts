
import { FontNames } from './constants'
import { Dimensions } from 'react-native'

export enum FontStyle {
    Regular = "Regular",
    Medium = "Medium",
    DemiBold = "DemiBold",
    Bold = "Bold",
}

export function avenirFont(style: FontStyle) {
    return FontNames.Avenir + "-" + style
}

export function windowWidth() {
    return Dimensions.get("window").width
}

export function windowHeight() {
    return Dimensions.get("window").height
}

export function parseNumberFromString(str: string) {
    return parseInt(str.replace(/,/g, ''), 10);
}