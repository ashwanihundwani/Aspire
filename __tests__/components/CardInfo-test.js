/**
 * @format
 */

import 'react-native';
import React from 'react';
import CardInfo from '../../src/components/CardInfo';
import renderer from 'react-test-renderer';
import { render, fireEvent, cleanup} from '@testing-library/react-native'
import { testIDs, DebitCardString } from '../../src/utils/constants';

it('renders correctly', () => {
    const tree = renderer
      .create(<CardInfo 
          holderName={"Ashwani"} 
          cardNumber="5000-7808-5006-2020" 
          expiry="12/22" 
          CVV="500"
          weeklySpentLimit="5,000"
          amountSpent="3,000"
          showProgress={true}/>)
      .toJSON();
    expect(tree).toMatchSnapshot();
});

it("shows/hides cvv on show/hide button click", () => {
  const { getByTestId, getByText, getAllByText } = render(<CardInfo 
    holderName={"Ashwani"} 
    cardNumber="5000-7808-5006-2020" 
    expiry="12/22" 
    CVV="500"
    weeklySpentLimit="5,000"
    amountSpent="3,000"
    showProgress={true}/>)

  // Check if Card Number & CVV hides
  fireEvent.press(getByTestId(testIDs.showHideButton))
  const { hiddenCVV, dotString } = DebitCardString.cardInfo
  getByText(hiddenCVV)
  getByText("2020")
  
  expect(getAllByText(dotString).length).toBe(3)

  // Check if Card Number & CVV shows up 
  fireEvent.press(getByTestId(testIDs.showHideButton))
  getByText("500") //cvv
  getByText("5000")
  getByText("7808")
  getByText("5006")
  getByText("2020")
})

afterAll(cleanup)

