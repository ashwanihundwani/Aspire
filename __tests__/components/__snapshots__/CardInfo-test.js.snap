// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "marginBottom": 34,
      "marginTop": 34,
      "width": "100%",
    }
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignSelf": "flex-end",
        "backgroundColor": "#FFFFFF",
        "borderTopLeftRadius": 6,
        "borderTopRightRadius": 6,
        "height": 38,
        "marginBottom": -10,
        "marginRight": 43.5,
        "opacity": 1,
        "width": 151,
      }
    }
    testID="DebitCard.CardInfo.showHide"
  >
    <View
      style={
        Object {
          "alignSelf": "center",
          "flexDirection": "row",
          "marginTop": 5,
        }
      }
    >
      <Image
        source={
          Object {
            "process": [Function],
          }
        }
      />
      <Text
        style={
          Object {
            "color": "#01D167",
            "fontFamily": "AvenirNext-DemiBold",
            "fontSize": 12,
            "marginLeft": 6,
          }
        }
      >
        Show card number
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "alignSelf": "center",
        "aspectRatio": 1.66,
        "backgroundColor": "#01D167",
        "borderRadius": 14,
        "marginTop": 0,
        "width": 663,
      }
    }
  >
    <Image
      source={
        Object {
          "process": [Function],
        }
      }
      style={
        Object {
          "alignSelf": "flex-end",
          "height": 21,
          "marginRight": 43.5,
          "marginTop": 43.5,
          "width": 74,
        }
      }
    />
    <Text
      style={
        Object {
          "color": "#FFFFFF",
          "fontFamily": "AvenirNext-Bold",
          "fontSize": 22,
          "marginLeft": 43.5,
          "marginTop": 43.5,
        }
      }
    >
      Ashwani
    </Text>
    <View
      style={
        Object {
          "flexDirection": "row",
          "marginLeft": 43.5,
          "marginTop": 43.5,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#FFFFFF",
            "fontFamily": "AvenirNext-DemiBold",
            "fontSize": 14,
            "letterSpacing": 3.46,
            "marginRight": 43.5,
          }
        }
      />
      <Text
        style={
          Object {
            "color": "#FFFFFF",
            "fontFamily": "AvenirNext-DemiBold",
            "fontSize": 14,
            "letterSpacing": 3.46,
            "marginRight": 43.5,
          }
        }
      />
      <Text
        style={
          Object {
            "color": "#FFFFFF",
            "fontFamily": "AvenirNext-DemiBold",
            "fontSize": 14,
            "letterSpacing": 3.46,
            "marginRight": 43.5,
          }
        }
      />
      <Text
        style={
          Object {
            "color": "#FFFFFF",
            "fontFamily": "AvenirNext-DemiBold",
            "fontSize": 14,
            "letterSpacing": 3.46,
            "marginRight": 43.5,
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "flex-start",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#FFFFFF",
            "fontFamily": "AvenirNext-DemiBold",
            "fontSize": 13,
            "marginLeft": 43.5,
            "marginTop": 15,
          }
        }
      >
        Thru: 12/22
      </Text>
      <Text
        style={
          Object {
            "color": "#FFFFFF",
            "fontFamily": "AvenirNext-DemiBold",
            "fontSize": 13,
            "marginLeft": 32,
            "marginTop": 15,
          }
        }
      >
        CVV: 
        <Text
          style={
            Object {
              "color": "#FFFFFF",
              "fontFamily": "AvenirNext-DemiBold",
              "fontSize": 13,
              "marginLeft": 32,
              "marginTop": 15,
            }
          }
        >
          500
        </Text>
      </Text>
    </View>
    <Image
      source={
        Object {
          "process": [Function],
        }
      }
      style={
        Object {
          "alignSelf": "flex-end",
          "height": 20,
          "marginRight": 43.5,
          "width": 59,
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "marginTop": 26,
        "paddingHorizontal": 43.5,
        "width": "100%",
      }
    }
    testID="DebitCard.SpendProgress"
  >
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginBottom": 6,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#222222",
            "fontFamily": "AvenirNext-Medium",
            "fontSize": 13,
          }
        }
      >
        Debit card spending limit
      </Text>
      <Text
        style={
          Object {
            "color": "#01D167",
            "fontFamily": "AvenirNext-DemiBold",
            "fontSize": 13,
          }
        }
      >
        $3,000
        <Text
          style={
            Object {
              "color": "#22222233",
              "fontFamily": "AvenirNext-DemiBold",
              "fontSize": 13,
            }
          }
        >
           | $5,000
        </Text>
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "backgroundColor": "rgba(1,209,103, 0.07)",
            "borderRadius": 15,
            "height": 15,
            "width": 663,
          },
          Object {
            "overflow": "hidden",
          },
        ]
      }
    >
      <View
        style={
          Object {
            "borderRightColor": "transparent",
            "borderRightWidth": 20,
            "borderTopColor": "#01D167",
            "borderTopWidth": 40,
            "height": 20,
            "width": 397.8,
          }
        }
      />
    </View>
  </View>
</View>
`;
